daemons/fenced/fenced_remote.c

```diff
@@ -2236,14 +2364,27 @@ fenced_process_fencing_reply(xmlNode *msg)
     }
 
     if (pcmk__str_eq(crm_element_value(msg, F_SUBTYPE), "broadcast", pcmk__str_casei)) {
-        if (pcmk__result_ok(&op->result)) {
-            op->state = st_done;
+        const char *type = NULL;
+
+        type = crm_element_value(msg, F_TYPE);
+
+        if (!pcmk__result_ok(&op->result) && pcmk__str_eq(type, T_STONITH_NG, pcmk__str_casei) &&
+            pcmk__str_eq(op->originator, op->target, pcmk__str_casei)) {
+            /* Handles escalation from fencing failures on DC nodes that do not have a topology. */
+            /* Only T_STONITH_NG is processed from send_async_reply(). */
+            if (!handles_dcnode_fencing_failures_no_topology(msg, op)) {
+                return;
+            }
```

Comment:
I'm having trouble following. Why is this checked for broadcast replies? I thought the goal was to send the broadcast ourselves if it doesn't arrive after doing send_async_reply().

Reply:
This is important.
If you track the response only on the node that broadcast the error, the response tracking will be lost, for example if this node goes down, and as a result the fencing operation will remain in the Pending state.
We have solved this problem by tracking the response on all nodes except the DC node.

Reply:
I guess what I don't understand is how it got broadcast to begin with. I thought the problem occurred when send_async_reply() sent the reply to only a single node, and that node did not send the broadcast.

Reply:
Hi Ken,

I'm sorry, I don't understand your question.
I've organized the issues and fixes.

The first problem is with multiple node configurations.
If there is no topology, fencing occurs due to a resource failure of the DC node, and if fencing from another node fails, this failure is notified by broadcast, so escalation fencing does not occur and fencing fails.

In order to avoid this, in the first PR, we proposed to avoid it by sending a notification of failure in this case to the DC node.
 - https://github.com/ClusterLabs/pacemaker/pull/2401

However, transmissions destined for non-broadcast DC nodes can fail, so this fencing failure can remain pending.

In order to avoid this, I thought about setting a timer to track the fencing operation after the node that fails in fencing sends. (When the timer fires, this node broadcasts the fencing failure to avoid becoming pending)

However, if the node that tracks the fencing operation here shuts down or causes the node to go down, it will be necessary to outsource the tracking of the fencing operation to other nodes in the cluster.
This outsourcing is not possible, especially if the node tracking the fencing operation goes down due to a failure.

For this reason, we decided to track fencing on all nodes by broadcasting the failure from the node where fencing failed.

If the error is successfully sent to the DC node, escalation fencing will be performed and even if the DC node is already down, the fencing operation history will change from pending to fail after the tracking timer is activated.
This history will then be successful later, with new DC nodes and planned fencing.

Does this explanation eliminate your doubts?

Best Regards,
Hideo Yamauchi.

Reply:
Hi Hideo,

Thanks for the explanation. If I follow correctly, this is the sequence of interest:

* The DC needs to fence itself, without topology, so it requests fencing as usual.
* Since the originator is also the target, the node that executes the fencing broadcasts the reply via send_async_reply().
* Every node processes the broadcast in fenced_process_fencing_reply(). The originator finalizes the op if successful and retries fencing if failed (this fixes the issue where fencing is not retried). All other nodes set a timer.
* If the timer expires before the op is finalized, then every node will finalize the op as failed (this avoids the op getting "stuck" as pending). The DC will reschedule fencing.

Is that correct?

Was there a reason that relaying non-topology self-fencing requests to another node (in check_alternate_host()) was not suitable?

Reply:
Hi Ken,

> Thanks for the explanation. If I follow correctly, this is the sequence of interest:
> 
> * The DC needs to fence itself, without topology, so it requests fencing as usual.
> * Since the originator is also the target, the node that executes the fencing broadcasts the reply via send_async_reply().
> * Every node processes the broadcast in fenced_process_fencing_reply(). The originator finalizes the op if successful and retries fencing if failed (this fixes the issue where fencing is not retried). All other nodes set a timer.
> * If the timer expires before the op is finalized, then every node will finalize the op as failed (this avoids the op getting "stuck" as pending). The DC will reschedule fencing.
> 
> Is that correct?

Yes.

> 
> Was there a reason that relaying non-topology self-fencing requests to another node (in check_alternate_host()) was not suitable?

I'll look into it a bit more, but at the time I was thinking about fixing it, it seemed fine to relay DC node failures even if there was no topology.

If you're okay with RELAYing, I'd rather RELAY than the fix I suggested.

What do you think?

Best Regards,
Hideo Yamauchi.

Reply:
> I'll look into it a bit more, but at the time I was thinking about fixing it, it seemed fine to relay DC node failures even if there was no topology.
> 
> If you're okay with RELAYing, I'd rather RELAY than the fix I suggested.
> 
> What do you think?

Hi Hideo,

I do think that would be better if it works, to avoid the possibility of timing issues and corner cases with synchronizing the pending replies.

Reply:
Hi Ken,

Okay!

I'll try a fix that makes DC fencing RELAY to other nodes even if there is no topology.
Also, if there is anything, I will consult with you.

Best Regards,
Hideo Yamauchi.

Reply:
Hi Ken,

Implemented PR to RELAY RELAY fencing for all DC failures.

 - https://github.com/ClusterLabs/pacemaker/pull/2815

Best Regards,
Hideo Yamauchi,